engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java

```diff
@@ -1068,26 +1095,104 @@ private static boolean isOperator(String txt, Either<MultipleOperatorError, Oper
+              case IR.Name name -> {
+                var type = translateQualifiedNameOrExpression(app.getRhs());
+                yield new IR$DefinitionArgument$Specified(
+                  name,
+                  Option.apply(type), Option.apply(defaultValue),
+                  false, getIdentifiedLocation(app), meta(), diag()
+                );
+              }
+              default -> throw new UnhandledEntity(app.getLhs(), "translateArgumentDefinition");
+            };
+          }
+          case Tree.TypeAnnotated anno -> {
+            yield null;
+          }
+          default -> throw new UnhandledEntity(withValue.getLhs(), "translateArgumentDefinition");
+        };
+      }
+      case Tree.OprSectionBoundary bound -> {
+          yield translateArgumentDefinition(bound.getAst(), null, isSuspended);
+      }
+      case Tree.TypeAnnotated anno -> {
+        yield null;
+      }
       case Tree.Ident id -> {
         IR.Expression identifier = translateIdent(id, false);
+        var defaultValue = translateExpression(defValue, false);
         yield switch (identifier) {
-          case IR.Name name_ -> name_;
-          // TODO: Other types of pattern. Needs IR support.
-          default -> throw new UnhandledEntity(pattern, "translateArgumentDefinition");
+          case IR.Name name -> new IR$DefinitionArgument$Specified(
+            name,
+            Option.empty(),
+            Option.apply(defaultValue),
+            isSuspended,
+            getIdentifiedLocation(arg),
+            meta(), diag()
+          );
+          default -> throw new UnhandledEntity(arg, "translateArgumentDefinition");
         };
       }
-      // TODO: Other types of pattern. Needs IR support.
-      default -> throw new UnhandledEntity(pattern, "translateArgumentDefinition");
+      case Tree.UnaryOprApp opr when "~".equals(opr.getOpr().codeRepr()) -> {
+        yield translateArgumentDefinition(opr.getRhs(), null, true);
+      }
+        /*
+      case AstView.AssignedArgument(name, value) =>
+        translateIdent(name) match {
+          case name: IR.Name =>
+            DefinitionArgument.Specified(
+              name,
+              None,
+              Some(translateExpression(value)),
+              isSuspended,
+              getIdentifiedLocation(arg)
+            )
+          case _ =>
+            throw new UnhandledEntity(arg, "translateArgumentDefinition")
+        }
+      */
+      case Tree.Wildcard wild -> {
+        var loc = getIdentifiedLocation(arg);
+        yield new IR$DefinitionArgument$Specified(
+            new IR$Name$Blank(loc, meta(), diag()),
+            Option.empty(),
+            Option.empty(),
+            isSuspended,
+            loc,
+            meta(), diag()
+          );
+      }
+      default -> throw new UnhandledEntity(core, "translateArgumentDefinition");
     };
-    boolean isSuspended = def.getSuspension() != null;
+    /*
```

Comment:
The old implementation was simpler. Is there a problem with it the tests don't catch?